x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java

```diff
@@ -1137,6 +1152,209 @@ protected Expression rule(Expression e) {
         }
     }
 
+    // Simplifies arithmetic expressions with fixed point fields in BinaryComparisons.
+    public static final class SimplifyComparisonsArithmetics extends OptimizerExpressionRule {
+        BiFunction<DataType, DataType, Boolean> typesCompatible;
+
+        public SimplifyComparisonsArithmetics(BiFunction<DataType, DataType, Boolean> typesCompatible) {
+            super(TransformDirection.UP);
+            this.typesCompatible = typesCompatible;
+        }
+
+        @Override
+        protected Expression rule(Expression e) {
+            return (e instanceof BinaryComparison) ? simplify((BinaryComparison) e) : e;
+        }
+
+        private Expression simplify(BinaryComparison bc) {
+            // optimize only once the expression has a literal on the right side of the binary comparison
+            if (bc.right() instanceof Literal) {
+                if (bc.left() instanceof ArithmeticOperation) {
+                    return SimplifyOperation.simplify(bc, typesCompatible);
+                } else if (bc.left() instanceof Neg) {
+                    return reduceNegation(bc);
+                }
+            }
+            return bc;
+        }
+
+        private static Expression reduceNegation(BinaryComparison bc) {
+            Literal bcLiteral = (Literal) bc.right();
+            Expression literalNeg = safeMaybeFold(new Neg(bcLiteral.source(), bcLiteral));
+            return literalNeg == null ? bc : bc.reverse().replaceChildren(List.of(((Neg) bc.left()).field(), literalNeg));
+        }
+
+        static Expression safeMaybeFold(Expression expression) {
```

Comment:
> I think the return null inside the catch block bothers me.

The `null` isn't returned from the catch block, it's just assigned to the var to return.
But this still being a detail, would you mind expanding what's troubling / how it could be improved?

> I would have kept the original (no separate method used)

The code is extracted as a method since it's reused. It should be acceptably concise too: "if the expression is foldable, attempt to do so. Return either folded outcome, original (if not foldable) or `null` (if foldable, but unsafely; i.e. exception wud be thrown)".
An alternative could be for the caller to save the input expression (as in your example), the function to return the original input, if folding is unsafe, and for caller to check returned value against the original input. But this would be more verbose (as in your example) and this pattern to be applied multiple times.

For now I've taken Costin's [suggestion](https://github.com/elastic/elasticsearch/pull/66022#discussion_r557683449), but still open for change if needed.

Reply:
Can be left as is I guess. My initial suggestion seems to be just a personal preference. If there will be other changes to this PR, I'd add a comment similar to the one I made in the `catch (ArithmeticException | DateTimeException e) {` to explain the catch block "ignoring" the math failure.